app/controllers/supplementary/ItemTypePageController.scala

```diff
@@ -42,21 +46,139 @@ class ItemTypePageController @Inject()(
+  ): Future[Result] = {
+    val itemTypeUpdated = updateCachedItemTypeAddition(itemTypeInput, itemTypeCache)
+    validateItemType(itemTypeUpdated).flatMap {
+      case Success =>
+        customsCacheService.cache[ItemType](supplementaryCacheId, ItemType.id, itemTypeUpdated).map { _ =>
+          Redirect(controllers.supplementary.routes.ItemTypePageController.displayPage())
         }
-      )
+      case Failure(errors) =>
+        val formWithErrors =
+          errors.foldLeft(ItemType.form.fill(itemTypeInput))((form, error) => form.withError(adjustErrorKey(error)))
+        Future.successful(
+          BadRequest(
+            item_type(
+              appConfig,
+              adjustDataKeys(formWithErrors),
+              itemTypeCache.taricAdditionalCodes,
+              itemTypeCache.nationalAdditionalCodes
+            )
+          )
+        )
+    }
   }
 
+  private def updateCachedItemTypeAddition(itemTypeInput: ItemType, itemTypeCache: ItemType): ItemType =
+    itemTypeCache.copy(
+      taricAdditionalCodes = itemTypeCache.taricAdditionalCodes ++ itemTypeInput.taricAdditionalCodes,
+      nationalAdditionalCodes = itemTypeCache.nationalAdditionalCodes ++ itemTypeInput.nationalAdditionalCodes
+    )
+
+  private def handleSaveAndContinue(itemTypeInput: ItemType, itemTypeCache: ItemType)(
+    implicit request: AuthenticatedRequest[AnyContent]
+  ): Future[Result] = {
+    val itemTypeUpdated = updateCachedItemTypeSaveAndContinue(itemTypeInput, itemTypeCache)
+    validateItemType(itemTypeUpdated).flatMap {
+      case Success =>
+        customsCacheService.cache[ItemType](supplementaryCacheId, ItemType.id, itemTypeUpdated).map { _ =>
+          Redirect(controllers.supplementary.routes.PackageInformationController.displayForm())
+        }
+      case Failure(errors) =>
+        val formWithErrors =
+          errors.foldLeft(ItemType.form.fill(itemTypeInput))((form, error) => form.withError(adjustErrorKey(error)))
+        Future.successful(
+          BadRequest(
+            item_type(
+              appConfig,
+              adjustDataKeys(formWithErrors),
+              itemTypeCache.taricAdditionalCodes,
+              itemTypeCache.nationalAdditionalCodes
+            )
+          )
+        )
+    }
+  }
+
+  private def updateCachedItemTypeSaveAndContinue(itemTypeInput: ItemType, itemTypeCache: ItemType): ItemType =
+    ItemType(
+      combinedNomenclatureCode = itemTypeInput.combinedNomenclatureCode,
+      taricAdditionalCodes = itemTypeCache.taricAdditionalCodes ++ itemTypeInput.taricAdditionalCodes,
+      nationalAdditionalCodes = itemTypeCache.nationalAdditionalCodes ++ itemTypeInput.nationalAdditionalCodes,
+      descriptionOfGoods = itemTypeInput.descriptionOfGoods,
+      cusCode = itemTypeInput.cusCode,
+      statisticalValue = itemTypeInput.statisticalValue
+    )
+
+  private def validateItemType(itemType: ItemType)(implicit request: Request[AnyContent]): Future[ValidationResult] =
```

Comment:
You can get rid of this method or pass action type. You use this method inside `handleAddition` and `handleSaveAndContinue` so you know what type you have. Based on it you don't need to use `getOrElse`

Reply:
Removed